packages/react/src/utils/useAnimationFrame.ts

```diff
@@ -0,0 +1,103 @@
+'use client';
+import { useLazyRef } from './useLazyRef';
+import { useOnMount } from './useOnMount';
+
+type AnimationFrameId = ReturnType<typeof requestAnimationFrame>;
+
+const EMPTY = 0 as unknown as AnimationFrameId;
+
+class Scheduler {
+  /* This implementation uses an array as a backing data-structure for frame callbacks.
+   * It allows `O(1)` callback cancelling by inserting a `null` in the array, though it
+   * never calls the native `cancelAnimationFrame` if there are no frames left. This can
+   * be much more efficient if there is a call pattern that alterns as
+   * "request-cancel-request-cancel-…".
+   * But in the case of "request-request-…-cancel-cancel-…", it leaves the final animation
+   * frame to run anyway. We turn that frame into a `O(1)` no-op via `callbacksCount`. */
```

Comment:
I only implemented the scheduler for animation frames, but if we have a situation where we call `setTimeout(*, N)` a lot where `N` is constant across calls during one tick of the event loop, we could also optimize that use-case to limit the number of `setTimeout` calls.